File: src/emulator.zig
const std = @import("std");
const fs = std.fs;
const time = std.time;

const CHIP8 = @import("chip8.zig").CHIP8;
const Platform = @import("platform.zig").Platform;

pub fn Emulator(comptime RenderType: type, comptime CHIP8Type: type, comptime DisplayBufferType: type) type {
    return struct {
        chip8: CHIP8Type,
        platform: Platform(RenderType, DisplayBufferType),
        /// The frame time in nanoseconds
        cycleDelay: u32,
        timer: time.Timer,

        const Self = @This();

        pub fn init(chip8: CHIP8Type, platform: Platform(RenderType, DisplayBufferType), cycleDelay: u32) !Self {
            var emulator = Self{
                .chip8 = chip8,
                .platform = platform,
                .cycleDelay = @floatToInt(u32, (1.0 / @intToFloat(f64, cycleDelay)) * 1_000_000_000), // *1_000_000_000: s to ns
                .timer = try time.Timer.start(),
            };

            std.debug.print("{}\n", .{emulator.cycleDelay});

            return emulator;
        }

        pub fn run(self: *Self) !void {
            self.timer.reset();

            var quit: bool = false;

            std.debug.print("running\n", .{});

            while (!quit) {
                quit = try self.platform.handleInput(self.platform.platform, &self.chip8.keypad);
                const dt: u64 = self.timer.read(); // / 1_000_000; // in nanoseconds to millis

                if (dt >= self.cycleDelay) {
                    // std.debug.print("{}\n", .{self.cycleDelay});
                    self.timer.reset();

                    self.chip8.cycle();
                    try self.platform.renderBuffer(self.platform.platform, &self.chip8.displayMemory);
                }
            }
        }

        pub fn loadROM(self: *Self, dir: fs.Dir, filename: []const u8) !void {
            try self.chip8.loadROM(dir, filename);
        }
    };
}

File: src/web/old/emulator.js
import { WASM } from './wasm.js';
import {
  renderBuffer,
  handleInput
} from './render.js';
import { Timer } from './timer.js';

export class Emulator {
  static async init(clockSpeed) {
    let emulator = new Emulator();
    emulator.wasm = await WASM.init("chip8-wasm.wasm");
    emulator.ptr = emulator.wasm.functions.initEmulator(clockSpeed);
    if (emulator.ptr == null) {
      console.error("Coudn't initialize emulator");
      return null;
    }
    emulator.worker = new Worker("wasm-worker.js");
    emulator.deinit = () => {
      emulator.wasm.functions.deinitEmulator(emulator.ptr);
    };
    emulator.loadROM = (rom) => {
      console.info("loading rom", rom);
      let memory = new Uint8Array(emulator.wasm.memory.buffer);
      const ptr = emulator.wasm.functions.alloc(rom.length);
      if (ptr == null) {
        console.error("Couldn't allocate rom data");
        return;
      }

      for (let i = 0; i < rom.length; i++) {
        memory[ptr + i] = rom[i];
      }
      
      if (emulator.wasm.functions.loadROM(emulator.ptr, 0x200, rom.length) == 1) {
        emulator.wasm.functions.free(ptr);
        throw "Error loading rom";
      }

      emulator.wasm.functions.free(ptr);
    };
    emulator.keypadPtr = emulator.wasm.functions.keypadPtr;
    emulator.handleInput = handleInput; // TODO
    emulator.timer = Timer.start();
    emulator.cycleDelay = (1.0 / clockSpeed) * 1_000_000_000;
    emulator.drawFlag = emulator.wasm.functions.drawFlag;
    emulator.displayMemPtr = emulator.wasm.functions.displayMemPtr;
    emulator.renderBuffer = renderBuffer;
    emulator.run = () => {
      const memory = new Uint8Array(emulator.wasm.memory.buffer);
      console.log("Running ROM");
      emulator.timer.reset();
      let quit = false;

      while (!quit) {
        quit = emulator.handleInput(emulator.keypadPtr);
        const dt = emulator.timer.read();

        if (dt >= emulator.cycleDelay) {
          emulator.timer.reset();

          emulator.tick();
          if (emulator.drawFlag() == 1) {
            emulator.renderBuffer(memory, emulator.displayMemPtr);
          }
        }
      }
      // if (emulator.wasm.functions.run(emulator.ptr) == 1) {
      //   throw "Error running emulator";
      // }
      
    };
    return emulator;
  }
}
